core_v14/src/decoder/decoder.rs

```diff
@@ -0,0 +1,196 @@
+        // we decode things from it.
+        let data = &mut data;
+
+		if data.len() == 0 {
+			return Err(DecodeError::EarlyEof("extrinsic length should be > 0"));
+		}
+
+        // V4 extrinsics (the format we can decode here) are laid out roughly as follows:
+		//
+		// first byte: abbbbbbb (a = 0 for unsigned, 1 for signed, b = version)
+        //
+		// signature, which is made up of (in order):
+		// - sp_runtime::MultiAddress enum (sender)
+		// - sp_runtime::MultiSignature enum
+		// - sp_runtime::generic::Era enum
+		// - compact encoded u32 (nonce; prior transaction count)
+		// - compact encoded u128 (tip paid to block producer/treasury)
+        //
+		// call, which is made up roughly of:
+		// - enum pallet index (for pallets variant)
+		// - call index (for inner variant)
+		// - call args (types can be pulled from metadata for each arg we expect)
+        //
+        // So, we start by getting the version/signed from the first byte and go from there.
+        let is_signed = data[0] & 0b1000_0000 != 0;
+        let version = data[0] & 0b0111_1111;
+        *data = &data[1..];
+
+		// We only know how to decode V4 extrinsics at the moment
+		if version != 4 {
+			return Err(DecodeError::CannotDecodeExtrinsicVersion(version));
+		}
+
+		// If the extrinsic is signed, decode the signature next.
+		let signature = match is_signed {
+            true => Some(decode_v4_signature(data, &self.metadata)?),
+            false => None
+        };
+
+        // Pluck out the u8's representing the pallet and call enum next.
+		if data.len() < 2 {
+			return Err(DecodeError::EarlyEof("expected at least 2 more bytes for the pallet/call index"));
+		}
+		let pallet_index = u8::decode(data)?;
+		let call_index = u8::decode(data)?;
+		log::trace!("pallet index: {}, call index: {}", pallet_index, call_index);
+
+		// Work out which call the extrinsic data represents and get type info for it:
+		let (pallet_name, call) = match self.metadata.call_by_variant_index(pallet_index, call_index) {
+			Some(call) => call,
+			None => return Err(DecodeError::CannotFindCall(pallet_index, call_index)),
+		};
+
+		// Decode each of the argument values in the extrinsic:
+		let mut arguments = vec![];
+		for arg in call.args() {
+			let ty = self.metadata.resolve_type(arg).ok_or(DecodeError::CannotFindType(arg.id()))?;
+			let val = match decode_type(data, &ty, &self.metadata) {
+				Ok(val) => val,
+				Err(err) => return Err(err.into()),
+			};
+			arguments.push(val);
+		}
+
+		let ext = GenericExtrinsic { pallet: pallet_name.to_owned(), call: call.name().to_owned(), signature, arguments };
+
+		// If there's data left to consume, it likely means we screwed up decoding:
+		if !data.is_empty() {
+			return Err(DecodeError::LateEof(*data, ext));
+		}
+
+		// Return a composite type representing the extrinsic arguments:
+		Ok(ext)
+	}
+}
+
+
+#[derive(Debug, Clone)]
+pub struct GenericExtrinsic {
```

Comment:
Would be good to derive `Serialize, Deserialize` for all the `Generic` types

Reply:
Hrm, `MultiAddress` doesn't impl (de)serialize :( Two options:
- impl Encode/Decode for everything and not impl (de)serialize,
- or define my own MultiAddress type which matches the shape of the substrate one and does impl those traits.

What do you reckon?